nsqd/nsqd.go

```diff
@@ -267,24 +268,45 @@ type meta struct {
 	} `json:"topics"`
 }
 
-func (n *NSQD) LoadMetadata() {
+func (n *NSQD) LoadMetadata() error {
 	atomic.StoreInt32(&n.isLoading, 1)
 	defer atomic.StoreInt32(&n.isLoading, 0)
 
-	fn := fmt.Sprintf(path.Join(n.getOpts().DataPath, "nsqd.%d.dat"), n.getOpts().ID)
+	fn := path.Join(n.getOpts().DataPath, "nsqd.dat")
+	// old metadata filename with ID, maintained in parallel to enable roll-back
+	fnID := fmt.Sprintf(path.Join(n.getOpts().DataPath, "nsqd.%d.dat"), n.getOpts().ID)
+
 	data, err := ioutil.ReadFile(fn)
 	if err != nil {
 		if !os.IsNotExist(err) {
-			n.logf("ERROR: failed to read channel metadata from %s - %s", fn, err)
+			return fmt.Errorf("failed to read channel metadata from %s - %s", fn, err)
 		}
-		return
+	}
+
+	dataID, errID := ioutil.ReadFile(fnID)
+	if errID != nil {
+		if !os.IsNotExist(errID) {
+			return fmt.Errorf("failed to read channel metadata from %s - %s", fnID, errID)
+		}
+	}
+
+	if err != nil && errID != nil {
+		return nil
+	}
+	if err == nil && errID == nil {
+		if bytes.Compare(data, dataID) != 0 {
+			return fmt.Errorf("metadata in %s and %s do not match (delete one)", fn, fnID)
+		}
+	}
+	if err != nil {
+		fn = fnID
+		data = dataID
```

Comment:
I've looked at this block multiple times and I still can't really make sense of it.

Is there a simpler way to implement this? What if `LoadMetadata` took a filename and the _caller_ did some of this compatibility logic?